metagraph/src/graph/graph_extensions/row_tuples_to_id.cpp

```diff
@@ -0,0 +1,123 @@
+#include "row_tuples_to_id.hpp"
+
+#include <sstream>
+
+#include <tsl/hopscotch_map.h>
+
+#include "common/serialization.hpp"
+#include "common/utils/file_utils.hpp"
+#include "common/threads/threading.hpp"
+
+namespace mtg::graph {
+
+
+RowTuplesToId::RowTuplesToId(const std::vector<std::string> &fai_infiles)
+      : seq_id_labels_(fai_infiles.size()), seq_delims_(fai_infiles.size()) {
+    size_t num_labels = fai_infiles.size();
+
+    #pragma omp parallel for schedule(dynamic)
+    for (size_t i = 0; i < num_labels; ++i) {
+        std::vector<uint64_t> delims;
+        std::ifstream fin(fai_infiles[i]);
+        std::string line;
+        std::string seq_id;
+        uint64_t len;
+        uint64_t cur_coord = 0;
+        while (std::getline(fin, line)) {
+            std::istringstream sin(line);
+            sin >> seq_id >> len;
+            cur_coord += len;
+            seq_id_labels_[i].emplace_back(std::move(seq_id));
+            delims.emplace_back(cur_coord);
+        }
+        if (delims.size()) {
+            seq_delims_[i] = bit_vector_sd([&](const auto &callback) {
+                for (uint64_t cur_coord : delims) {
+                    callback(cur_coord - 1);
+                }
+            }, delims.back(), delims.size());
+        }
+    }
+}
+
+std::vector<std::tuple<std::string, size_t, std::vector<SmallVector<uint64_t>>>> RowTuplesToId
+::rows_tuples_to_label_tuples(const std::vector<RowTuples> &rows_tuples) const {
+    using KmerCoords = std::vector<SmallVector<uint64_t>>;
+    tsl::hopscotch_map<std::uint64_t, std::tuple<std::string, size_t, KmerCoords>> conv_coords(seq_id_labels_.size());
+    for (size_t i = 0; i < rows_tuples.size(); ++i) {
+        const auto &row_tuples = rows_tuples[i];
+        for (const auto &[col, tuple] : row_tuples) {
+            assert(col < seq_id_labels_.size());
+            assert(col < seq_delims_.size());
+            const auto &labels = seq_id_labels_[col];
+            const auto &delims = seq_delims_[col];
+            for (uint64_t coord : tuple) {
+                uint64_t seq_id = delims.rank1(coord);
+                uint64_t conv_coord = seq_id > 0 ? coord - delims.select1(seq_id) : coord;
```

Comment:
This can be optimized. Often coordinates are consequitive, so we can reuse previously computed values